### DISCUSSION TITLE
Human limbs tracking with XSens MTw Awinda IMUs

### DISCUSSION THREAD

**Discussion by torydebra** [robotology/human-dynamics-estimation#228] [Ideas]
Hi, Internet bring me here

I am Davide, an IIT PHD student, working on teleoperation. I get in contact with some of you by email.

I am looking for a framework to perform some human motion tracking, and I was wondering about the feasibility of doing this with your framework. 
What we have are 6x XSens MTw Awinda Trackers ([link](https://shop.xsens.com/shop/xsens-mvn/mtw-awinda-wireless-3dof-motion-tracker)). As you might know, the XSens firmware already provides time syncronization and data from each sensors (angle (euler/quaternion), angular velocity and acceleration, magnetic field). 

What I am looking for is a way to combine data from the multiple devices attached to human body to track human motions. We would like to teleoperate the robot beginning from the operator end-effector position-velocity-acceleration (still thinking if all of these are necessary). We are starting with a single operator arm, but we want soon to use also the other arm (3 trackers for arm?). In future we may enhance using also the legs.

XSens software, MVN Analyze ([link](https://www.xsens.com/products/mvn-analyze)) but it is only available for Windows, and we would like to run everything on a single Linux (Ubuntu) machine.
I saw that your software already support XSens suit and has interface for ROS, things that we are looking for.

Thanks for your work and for any support!

---
fyi @liesrock @ntsagarakis

**Comment by yeshasvitirupachuri** [comment]
CC @lrapetti @claudia-lat @S-Dafarra @traversaro @DanielePucci

**Comment by yeshasvitirupachuri** [comment]
Hi @torydebra 

Thanks for opening this discussion.

 > I am looking for a framework to perform some human motion tracking, and I was wondering about the feasibility of doing this with your framework.

Let me give a brief overview of our infrastructure that can hopefully help you to understand how it can be useful for your purpose. The [wearables library](https://github.com/robotology/wearables) and [human-dynamics-estimation](https://github.com/robotology/human-dynamics-estimation) project are the two main components of the infrastructure.

### Wearables library   

Thanks to the work done by @diegoferigo and @lucaTagliapietra , the wearables library provides [general sensor interfaces](https://github.com/robotology/wearables/tree/master/interfaces/IWear/include/Wearable/IWear/Sensors) that can be implemented to represent a set or a combination of sensors that constitute a wearable device. An example of such a wearable device is the [XsensSuit](https://github.com/robotology/wearables/tree/master/devices/XsensSuit).  We have been primarily working with whole-body suit of xsens i.e., xsens Link (wired system) and xsens Awinda  (wireless system) that constitute 17 distributed IMUs ([Reference](https://www.xsens.com/products/mvn-animate)). However, instead of using Xsens MVN Motion Capture Studio to setup, configure and run the whole-body suit, we opted to use xsens sdk (Version 2018.03)  and developed [XSensMVN](https://github.com/robotology/wearables/tree/master/XSensMVN) library to perform the setup, configuration and accessing the sensor data from the system. 

The [XsensSuit](https://github.com/robotology/wearables/tree/master/devices/XsensSuit) wearable device accesses various sensors data through  [XSensMVN](https://github.com/robotology/wearables/tree/master/XSensMVN) library, and [implements the relevant sensor interfaces](https://github.com/robotology/wearables/blob/master/devices/XsensSuit/include/XsensSuit.h#L88-L110) of the wearable library.  The encapsulated sensors data from the `XsensSuit` wearable device is exposed through the [IWear interface](https://github.com/robotology/wearables/blob/master/interfaces/IWear/include/Wearable/IWear/IWear.h) which represents a unique data format referred to as `wearable data`. The wearable data from the particular wearable devices are streamed to **yarp network** in particular through [IWearWrapper](https://github.com/robotology/wearables/tree/master/wrappers/IWear).

An example of a single wearable data point from the `XsensSuit` is a following:

```
0 2406122.277751 1600691785.939639 "XsensSuit::" () () () () (("XsensSuit::fbAcc::Head" ("XsensSuit::fbAcc::Head" 1 0.0236732254672339728629 0.0305110471518924009082 0.000796472539171588778117)) ("XsensSuit::fbAcc::LeftFoot" ("XsensSuit::fbAcc::LeftFoot" 1 -0.0323633518985322332639 0.0485243470674375301788 0.00901567588726592816784)) ("XsensSuit::fbAcc::LeftForeArm" ("XsensSuit::fbAcc::LeftForeArm" 1 0.0494507328091359221744 0.0885437251161213523254 0.0024061895823184187293)) ("XsensSuit::fbAcc::LeftHand" ("XsensSuit::fbAcc::LeftHand" 1 0.0561218192473700305811 0.0353489857872835849806 -0.0152144913066578624239)) ("XsensSuit::fbAcc::LeftLowerLeg" ("XsensSuit::fbAcc::LeftLowerLeg" 1 0.00129620339565407471927 0.0649806845864996096562 -0.00317983682323763616751)) ("XsensSuit::fbAcc::LeftShoulder" ("XsensSuit::fbAcc::LeftShoulder" 1 0.0116267817264053413551 0.0646367817193072013993 0.011927396452421135109)) ("XsensSuit::fbAcc::LeftUpperArm" ("XsensSuit::fbAcc::LeftUpperArm" 1 0.00873818696930456217942 0.0592709573434314768159 -0.000406393929443638057819)) ("XsensSuit::fbAcc::LeftUpperLeg" ("XsensSuit::fbAcc::LeftUpperLeg" 1 -0.0131558453119839935397 0.0317034635285865645793 0.00784832628040810220682)) ("XsensSuit::fbAcc::Pelvis" ("XsensSuit::fbAcc::Pelvis" 1 -0.00993373095955957241543 0.0422665536003986874736 0.00458819324454573518324)) ("XsensSuit::fbAcc::RightFoot" ("XsensSuit::fbAcc::RightFoot" 1 -0.0394687692219919811221 0.0414373382102130316529 -0.00440693805453484799273)) ("XsensSuit::fbAcc::RightForeArm" ("XsensSuit::fbAcc::RightForeArm" 1 -0.0280346088966480708193 0.0443812499109758409888 -0.00889529178999204361844)) ("XsensSuit::fbAcc::RightHand" ("XsensSuit::fbAcc::RightHand" 1 -0.0197263579656027296383 0.0281972502830642873795 -0.0171736081383624394903)) ("XsensSuit::fbAcc::RightLowerLeg" ("XsensSuit::fbAcc::RightLowerLeg" 1 -0.00789535072719588226775 0.046827450646058504935 -0.00935629618092548298591)) ("XsensSuit::fbAcc::RightShoulder" ("XsensSuit::fbAcc::RightShoulder" 1 0.0222178740439035911358 0.0418325929348976471855 0.00653049935691996097376)) ("XsensSuit::fbAcc::RightUpperArm" ("XsensSuit::fbAcc::RightUpperArm" 1 -0.0245211833693058413319 0.0166426851772863981793 -0.0266979705387395682692)) ("XsensSuit::fbAcc::RightUpperLeg" ("XsensSuit::fbAcc::RightUpperLeg" 1 -0.00557677611392877903979 0.0578502440027625491892 -0.00109769836083195571064)) ("XsensSuit::fbAcc::T8" ("XsensSuit::fbAcc::T8" 1 0.0567773227331463958834 0.0809086313541468316046 0.00186343615862227807156))) () (("XsensSuit::mag::Head" ("XsensSuit::mag::Head" 1 -0.094482421875 -0.773681640625 0.2490234375)) ("XsensSuit::mag::LeftFoot" ("XsensSuit::mag::LeftFoot" 1 -0.60009765625 -0.275146484375 -0.490966796875)) ("XsensSuit::mag::LeftForeArm" ("XsensSuit::mag::LeftForeArm" 1 -0.794189453125 -0.262451171875 0.171142578125)) ("XsensSuit::mag::LeftHand" ("XsensSuit::mag::LeftHand" 1 -0.797119140625 0.04638671875 0.3095703125)) ("XsensSuit::mag::LeftLowerLeg" ("XsensSuit::mag::LeftLowerLeg" 1 -0.822998046875 -0.26318359375 0.337158203125)) ("XsensSuit::mag::LeftShoulder" ("XsensSuit::mag::LeftShoulder" 1 -0.262939453125 -0.738037109375 -0.28125)) ("XsensSuit::mag::LeftUpperArm" ("XsensSuit::mag::LeftUpperArm" 1 -0.785400390625 -0.07080078125 0.2939453125)) ("XsensSuit::mag::LeftUpperLeg" ("XsensSuit::mag::LeftUpperLeg" 1 -0.82568359375 0.19775390625 0.255859375)) ("XsensSuit::mag::Pelvis" ("XsensSuit::mag::Pelvis" 1 -0.818359375 0.2861328125 -0.05908203125)) ("XsensSuit::mag::RightFoot" ("XsensSuit::mag::RightFoot" 1 -0.676025390625 0.1982421875 -0.5908203125)) ("XsensSuit::mag::RightForeArm" ("XsensSuit::mag::RightForeArm" 1 -0.75732421875 -0.143798828125 -0.4091796875)) ("XsensSuit::mag::RightHand" ("XsensSuit::mag::RightHand" 1 -0.74365234375 -0.106689453125 -0.21826171875)) ("XsensSuit::mag::RightLowerLeg" ("XsensSuit::mag::RightLowerLeg" 1 -0.798095703125 -0.289306640625 0.468505859375)) ("XsensSuit::mag::RightShoulder" ("XsensSuit::mag::RightShoulder" 1 0.235107421875 0.6865234375 -0.3173828125)) ("XsensSuit::mag::RightUpperArm" ("XsensSuit::mag::RightUpperArm" 1 -0.811767578125 0.041259765625 -0.353515625)) ("XsensSuit::mag::RightUpperLeg" ("XsensSuit::mag::RightUpperLeg" 1 -0.99072265625 -0.031494140625 -0.19384765625)) ("XsensSuit::mag::T8" ("XsensSuit::mag::T8" 1 -0.73583984375 -0.196044921875 -0.351318359375))) (("XsensSuit::orient::Head" ("XsensSuit::orient::Head" 1 0.645082079310157485708 0.728720604686428852226 -0.0867823978723567535098 -0.212847848656859489624)) ("XsensSuit::orient::LeftFoot" ("XsensSuit::orient::LeftFoot" 1 0.693685952299794905684 -0.118094084657068429389 -0.286684630076847579705 -0.650127302632829917606)) ("XsensSuit::orient::LeftForeArm" ("XsensSuit::orient::LeftForeArm" 1 0.740542994193217452015 -0.0686095043892901157445 -0.647660420519582058496 -0.165604315617681130401)) ("XsensSuit::orient::LeftHand" ("XsensSuit::orient::LeftHand" 1 0.691810865654365891864 -0.185177442777027106979 -0.696037079696852245547 -0.0513753300349175726081)) ("XsensSuit::orient::LeftLowerLeg" ("XsensSuit::orient::LeftLowerLeg" 1 0.34273811195810832686 -0.699242846859586042463 -0.413045370389213017948 -0.4722113401085681339)) ("XsensSuit::orient::LeftShoulder" ("XsensSuit::orient::LeftShoulder" 1 0.857380191850293349454 0.483953483349952973125 -0.0817767707407091870575 -0.154921891104592035493)) ("XsensSuit::orient::LeftUpperArm" ("XsensSuit::orient::LeftUpperArm" 1 0.714744979774251443239 -0.0648462046062547037284 -0.694833536128865003612 -0.0462703004780279439356)) ("XsensSuit::orient::LeftUpperLeg" ("XsensSuit::orient::LeftUpperLeg" 1 0.679291530857277114919 0.0110738649006844012967 -0.732500890949520311324 0.0433915934242996292891)) ("XsensSuit::orient::Pelvis" ("XsensSuit::orient::Pelvis" 1 0.576928832693512139151 0.434664637544853571516 -0.505716068010364838692 0.471668351101747318488)) ("XsensSuit::orient::RightFoot" ("XsensSuit::orient::RightFoot" 1 0.371067961650336120005 -0.365890885666409704235 0.0315457631561619766813 -0.852899344852493634939)) ("XsensSuit::orient::RightForeArm" ("XsensSuit::orient::RightForeArm" 1 0.147059326620592206725 -0.63135846178133714357 -0.0711054747757902116945 -0.758092381341425869401)) ("XsensSuit::orient::RightHand" ("XsensSuit::orient::RightHand" 1 0.168362581429441243719 -0.686044491634640762889 -0.023232407437362028596 -0.707430033230741162065)) ("XsensSuit::orient::RightLowerLeg" ("XsensSuit::orient::RightLowerLeg" 1 0.456005574040443728556 -0.47105003616855661841 -0.641893175646423785885 -0.397673145847438236355)) ("XsensSuit::orient::RightShoulder" ("XsensSuit::orient::RightShoulder" 1 0.0730252954756389904833 0.0191690049339720385013 -0.506892149023613569625 0.858696806054827344035)) ("XsensSuit::orient::RightUpperArm" ("XsensSuit::orient::RightUpperArm" 1 0.00137960919170445825421 -0.694110478056340518549 0.00959462013844092874437 -0.719803226024499642755)) ("XsensSuit::orient::RightUpperLeg" ("XsensSuit::orient::RightUpperLeg" 1 0.0568104018999197116213 0.744029774377916397476 -0.00455399704685346094768 0.665711299427922398309)) ("XsensSuit::orient::T8" ("XsensSuit::orient::T8" 1 0.583741461099714475935 -0.481404930338703629378 -0.309061312564717693974 -0.576173849384483682812))) (("XsensSuit::pose::Head" ("XsensSuit::pose::Head" 1 0.645082079310157485708 0.728720604686428852226 -0.0867823978723567535098 -0.212847848656859489624 0.269285760201929680324 -1.53036340898865796945 1.60727348668370506601)) ("XsensSuit::pose::LeftFoot" ("XsensSuit::pose::LeftFoot" 1 0.693685952299794905684 -0.118094084657068429389 -0.286684630076847579705 -0.650127302632829917606 0.145409438392597734069 -1.47138076706834897145 0.0831502423504998366965)) ("XsensSuit::pose::LeftForeArm" ("XsensSuit::pose::LeftForeArm" 1 0.740542994193217452015 -0.0686095043892901157445 -0.647660420519582058496 -0.165604315617681130401 0.106913559145318193266 -1.56031329359853332406 0.894607867557315472595)) ("XsensSuit::pose::LeftHand" ("XsensSuit::pose::LeftHand" 1 0.691810865654365891864 -0.185177442777027106979 -0.696037079696852245547 -0.0513753300349175726081 0.128498470410860488045 -1.58185271622726641993 0.800922867563036544603)) ("XsensSuit::pose::LeftLowerLeg" ("XsensSuit::pose::LeftLowerLeg" 1 0.34273811195810832686 -0.699242846859586042463 -0.413045370389213017948 -0.4722113401085681339 0.169553394999319040348 -1.47270400985275684569 0.355941474534118784678)) ("XsensSuit::pose::LeftShoulder" ("XsensSuit::pose::LeftShoulder" 1 0.857380191850293349454 0.483953483349952973125 -0.0817767707407091870575 -0.154921891104592035493 0.199283266398917052742 -1.56111884837977310525 1.32856963001327987861)) ("XsensSuit::pose::LeftUpperArm" ("XsensSuit::pose::LeftUpperArm" 1 0.714744979774251443239 -0.0648462046062547037284 -0.694833536128865003612 -0.0462703004780279439356 0.0610364399693719000828 -1.54477638112720416785 1.27465872077746400493)) ("XsensSuit::pose::LeftUpperLeg" ("XsensSuit::pose::LeftUpperLeg" 1 0.679291530857277114919 0.0110738649006844012967 -0.732500890949520311324 0.0433915934242996292891 0.129596669811513215409 -1.49109591948658226634 0.67705348803136466973)) ("XsensSuit::pose::Pelvis" ("XsensSuit::pose::Pelvis" 1 0.576928832693512139151 0.434664637544853571516 -0.505716068010364838692 0.471668351101747318488 0.264432015719247737184 -1.59865432864239731536 1.02711071035345802294)) ("XsensSuit::pose::RightFoot" ("XsensSuit::pose::RightFoot" 1 0.371067961650336120005 -0.365890885666409704235 0.0315457631561619766813 -0.852899344852493634939 0.353833737845105633646 -1.48705557465830562336 0.0903012708093731686354)) ("XsensSuit::pose::RightForeArm" ("XsensSuit::pose::RightForeArm" 1 0.147059326620592206725 -0.63135846178133714357 -0.0711054747757902116945 -0.758092381341425869401 0.429785543570702033289 -1.58422826298561503222 0.898346603730311166203)) ("XsensSuit::pose::RightHand" ("XsensSuit::pose::RightHand" 1 0.168362581429441243719 -0.686044491634640762889 -0.023232407437362028596 -0.707430033230741162065 0.460843065092828463314 -1.58688718503655246117 0.802258842323521292883)) ("XsensSuit::pose::RightLowerLeg" ("XsensSuit::pose::RightLowerLeg" 1 0.456005574040443728556 -0.47105003616855661841 -0.641893175646423785885 -0.397673145847438236355 0.351805084993229655321 -1.5220353026784647632 0.359543946163210115241)) ("XsensSuit::pose::RightShoulder" ("XsensSuit::pose::RightShoulder" 1 0.0730252954756389904833 0.0191690049339720385013 -0.506892149023613569625 0.858696806054827344035 0.332742797192620176627 -1.58667433027497928322 1.33332864729439282137)) ("XsensSuit::pose::RightUpperArm" ("XsensSuit::pose::RightUpperArm" 1 0.00137960919170445825421 -0.694110478056340518549 0.00959462013844092874437 -0.719803226024499642755 0.444342593141102304433 -1.611044138574352802 1.28211804640885063122)) ("XsensSuit::pose::RightUpperLeg" ("XsensSuit::pose::RightUpperLeg" 1 0.0568104018999197116213 0.744029774377916397476 -0.00455399704685346094768 0.665711299427922398309 0.423460927984600332685 -1.50221875311433583633 0.67328870217261815867)) ("XsensSuit::pose::T8" ("XsensSuit::pose::T8" 1 0.583741461099714475935 -0.481404930338703629378 -0.309061312564717693974 -0.576173849384483682812 0.311853649093316531715 -1.37938604327894775992 1.38543999785938076741))) (("XsensSuit::pos::Head" ("XsensSuit::pos::Head" 1 0.269285760201929680324 -1.53036340898865796945 1.60727348668370506601)) ("XsensSuit::pos::LeftFoot" ("XsensSuit::pos::LeftFoot" 1 0.145409438392597734069 -1.47138076706834897145 0.0831502423504998366965)) ("XsensSuit::pos::LeftForeArm" ("XsensSuit::pos::LeftForeArm" 1 0.106913559145318193266 -1.56031329359853332406 0.894607867557315472595)) ("XsensSuit::pos::LeftHand" ("XsensSuit::pos::LeftHand" 1 0.128498470410860488045 -1.58185271622726641993 0.800922867563036544603)) ("XsensSuit::pos::LeftLowerLeg" ("XsensSuit::pos::LeftLowerLeg" 1 0.169553394999319040348 -1.47270400985275684569 0.355941474534118784678)) ("XsensSuit::pos::LeftShoulder" ("XsensSuit::pos::LeftShoulder" 1 0.199283266398917052742 -1.56111884837977310525 1.32856963001327987861)) ("XsensSuit::pos::LeftUpperArm" ("XsensSuit::pos::LeftUpperArm" 1 0.0610364399693719000828 -1.54477638112720416785 1.27465872077746400493)) ("XsensSuit::pos::LeftUpperLeg" ("XsensSuit::pos::LeftUpperLeg" 1 0.129596669811513215409 -1.49109591948658226634 0.67705348803136466973)) ("XsensSuit::pos::Pelvis" ("XsensSuit::pos::Pelvis" 1 0.264432015719247737184 -1.59865432864239731536 1.02711071035345802294)) ("XsensSuit::pos::RightFoot" ("XsensSuit::pos::RightFoot" 1 0.353833737845105633646 -1.48705557465830562336 0.0903012708093731686354)) ("XsensSuit::pos::RightForeArm" ("XsensSuit::pos::RightForeArm" 1 0.429785543570702033289 -1.58422826298561503222 0.898346603730311166203)) ("XsensSuit::pos::RightHand" ("XsensSuit::pos::RightHand" 1 0.460843065092828463314 -1.58688718503655246117 0.802258842323521292883)) ("XsensSuit::pos::RightLowerLeg" ("XsensSuit::pos::RightLowerLeg" 1 0.351805084993229655321 -1.5220353026784647632 0.359543946163210115241)) ("XsensSuit::pos::RightShoulder" ("XsensSuit::pos::RightShoulder" 1 0.332742797192620176627 -1.58667433027497928322 1.33332864729439282137))
```

So, to summarize in bullet points:

- [XSensMVN](https://github.com/robotology/wearables/tree/master/XSensMVN) library uses the xsens sdk to configure the system and access sensor data
-  [XsensSuit](https://github.com/robotology/wearables/tree/master/devices/XsensSuit) is a wearable device implementation using the sensor interfaces of wearables library. It accesses data from xsens sensors through [XSensMVN](https://github.com/robotology/wearables/tree/master/XSensMVN) library 
- [IWearWrapper](https://github.com/robotology/wearables/tree/master/wrappers/IWear) exposes the `wearable data` of the [XsensSuit](https://github.com/robotology/wearables/tree/master/devices/XsensSuit) wearable device to the **YARP network**

The different devices involved in this pipeline are launch through this [yarprobotinterface  configuration file](https://github.com/robotology/wearables/blob/master/app/xml/XsensSuitWearableDevice.xml)


We refer to this aspect of our infrastructure as the **PRODUCER** layer as indicated in the [schema picture](https://raw.githubusercontent.com/robotology/human-dynamics-estimation/master/misc/hde_scheme.png). The reason being, this layer consists of all the components that **produce** the `wearable data` from different wearable devices. 

**NOTE:** We run this part of the infrastructure on a windows machine as the xsens SDK is not supported on a Ubuntu machine

### Human Dynamics Estimation

We refer to this aspect of our infrastructure as the  **CONSUMER** layer as indicated in the [schema picture](https://raw.githubusercontent.com/robotology/human-dynamics-estimation/master/misc/hde_scheme.png). The reason being, this layer consists of all the components that **consume** the `wearable data` from **PRODUCER** layer.

The first device at the **CONSUMER** layer is the [IWearRemapper](https://github.com/robotology/wearables/tree/master/devices/IWearRemapper) that reads the `wearable data` from the YARP network, and exposes the different sensors data through `IWear` interface. This is a way to unpack/deserialize the data coming from the wearable device to the corresponding  sensor data. 

At this layer we organize the components mainly into three categories:

- Computational Units: Devices that access the sensor data through from the `IWearRemapper` using `IWear` interface. An example of such a computational unit is the [Human State Provider](https://github.com/robotology/human-dynamics-estimation/tree/master/devices/HumanStateProvider), which reads the xsens sensor data and performs inverse kinematics to estimate the state information. The data from a computational unit is typically exposed through dedicated interfaces e.g. [IHumanState](https://github.com/robotology/human-dynamics-estimation/tree/master/interfaces/IHumanState) interface
- Publishers: Devices that access information from computational units through the corresponding interface and exposes desired data to the ROS network. An example is [HumanStatePublisher](https://github.com/robotology/human-dynamics-estimation/tree/master/publishers/HumanStatePublisher)
- Wrappers: Devices that access information from the computational units through the corresponding interface and exposes desired data to the YARP network. An example is [HumanStateWrapper](https://github.com/robotology/human-dynamics-estimation/tree/master/wrappers/HumanStateWrapper) 

Similar to the producer layer, the different devices involved in this consumer layer are launch through a [yarprobotinterface  configuration file](https://github.com/robotology/human-dynamics-estimation/blob/master/conf/xml/HumanStateProvider.xml)

> What I am looking for is a way to combine data from the multiple devices attached to human body to track human motions. 

I hope the above description gives a better picture of our infrastructure. Having said that, the first check that can be very useful is to see how your hardware can be configured through xsens SDK.  In our case when we use the whole-body xsens awinda suit we use [these configuration options](https://github.com/robotology/wearables/blob/master/app/xml/XsensSuitWearableDevice.xml#L6-L48).  Once you can identify that, we can start a discussion on what modifications are needed on [XSensMVN](https://github.com/robotology/wearables/tree/master/XSensMVN) library to access the data from the sensors.

**Comment by torydebra** [comment]
Thanks @Yeshasvitvs for your overview. Things are becoming clear.

I will list some questions:

1.  > NOTE: We run this part of the infrastructure on a windows machine as the xsens SDK is not supported on a Ubuntu machine

  I am not sure about this. What is exactly the SDK are you speaking about ? If it is the firmware+software that works as a driver for each device, and provides the data to the system, this **is** available on Linux, at least for the Awindas. You can see it in the [download](https://www.xsens.com/software-downloads) page under the Awinda box, [here](https://www.xsens.com/hubfs/Downloads/Software/MTSS/Releases/4.6.0/MT%20Software%20Suite%204.6%20MTw%20Release%20Notes.pdf) the release notes.


2. I see on [yarprobotinterface configuration file](https://github.com/robotology/wearables/blob/master/app/xml/XsensSuitWearableDevice.xml) the lines:
~~~xml
        <!-- XsensSuit configuration to be used. Available values are:-->
        <!-- FullBody -  FullBodyNoSternum - FullBodyNoHands - FullBodyNoSternumNoHands -->
        <!-- LowerBody - UpperBody - UpperBodyNoSternum - UpperBodyNoHands - UpperBodyNoSternumNoHands -->
        <!-- FullBodyNoShoulders - SingleDevice - PelvisSternum - LowerBodyPlus -->
        <param name="suit-config">FullBody</param>
~~~
  What exactly is each configuration? How many sensors each configurations need? Is there a configuration that uses 3 Awinda IMU for each arm, to track only these arms? Which is the minimum number of sensors to let the Human State Provider tracks an arm?

3. Is the calibration (N/T pose and so on) performed by the Producer or the Consumer?

4. What I am thinking "a caldo" is to provide MTW Awinda data + configuration/calibration to the IWearRemapper directly, without passing through the network. If what I am saying in 1 is correct, this would permit to run everything on a single (Linux) machine. If I have explained myself well, is this feasible for you?

**Reply by traversaro** [reply]
>     1. > NOTE: We run this part of the infrastructure on a windows machine as the xsens SDK is not supported on a Ubuntu machine
> 
> 
> I am not sure about this. What is exactly the SDK are you speaking about ? If it is the firmware+software that works as a driver for each device, and provides the data to the system, this **is** available on Linux, at least for the Awindas. You can see it in the [download](https://www.xsens.com/software-downloads) page under the Awinda box, [here](https://www.xsens.com/hubfs/Downloads/Software/MTSS/Releases/4.6.0/MT%20Software%20Suite%204.6%20MTw%20Release%20Notes.pdf) the release notes.

I think that the wearables repo is using the so-called "Xsens MVN SDK" that exposes the XME API, that is now discontinued by Xsens, but probably you can find if from the old downloads on the Xsens website or asking to Xsens, probably @Yeshasvitvs or @lrapetti can know more about this. That was a SDK that as far as I know is different from the MTw, and is only running on Windows. For more info, feel free to check the code in https://github.com/robotology/wearables/blob/master/XSensMVN/src/XSensMVNDriverImpl.cpp .


>     I see on [yarprobotinterface configuration file](https://github.com/robotology/wearables/blob/master/app/xml/XsensSuitWearableDevice.xml) the lines:
> 
> 
> ```
>         
>         
>         
>         
>         <param name="suit-config">FullBody</param>
> ```
> 
> What exactly is each configuration? How many sensors each configurations need? Is there a configuration that uses 3 Awinda IMU for each arm, to track only these arms? Which is the minimum number of sensors to let the Human State Provider tracks an arm?

The code is open source, so feel free to check it! In particular, that parameters is read in https://github.com/robotology/wearables/blob/77362fc692bb9549b168b10c1200c473637bd59e/devices/XsensSuit/src/XsensSuit.cpp#L661 and passed to the Xsens SDK in https://github.com/robotology/wearables/blob/77362fc692bb9549b168b10c1200c473637bd59e/devices/XsensSuit/src/XsensSuit.cpp#L777 .

**Comment by traversaro** [comment]
>    > NOTE: We run this part of the infrastructure on a windows machine as the xsens SDK is not supported on a Ubuntu machine
> 
> 
> I am not sure about this. What is exactly the SDK are you speaking about ? If it is the firmware+software that works as a driver for each device, and provides the data to the system, this **is** available on Linux, at least for the Awindas. You can see it in the [download](https://www.xsens.com/software-downloads) page under the Awinda box, [here](https://www.xsens.com/hubfs/Downloads/Software/MTSS/Releases/4.6.0/MT%20Software%20Suite%204.6%20MTw%20Release%20Notes.pdf) the release notes.

I think that the wearables repo is using the so-called "Xsens MVN SDK" that exposes the XME API, that is now discontinued by Xsens, but probably you can find if from the old downloads on the Xsens website or asking to Xsens, probably @Yeshasvitvs or @lrapetti can know more about this. That was a SDK that as far as I know is different from the MTw, and is only running on Windows. For more info, feel free to check the code in https://github.com/robotology/wearables/blob/master/XSensMVN/src/XSensMVNDriverImpl.cpp .


>  I see on [yarprobotinterface configuration file](https://github.com/robotology/wearables/blob/master/app/xml/XsensSuitWearableDevice.xml) the lines:
> 
> 
> ```
>         
>         
>         
>         
>         <param name="suit-config">FullBody</param>
> ```
> 
> What exactly is each configuration? How many sensors each configurations need? Is there a configuration that uses 3 Awinda IMU for each arm, to track only these arms? Which is the minimum number of sensors to let the Human State Provider tracks an arm?

The code is open source, so feel free to check it! In particular, that parameters is read in https://github.com/robotology/wearables/blob/77362fc692bb9549b168b10c1200c473637bd59e/devices/XsensSuit/src/XsensSuit.cpp#L661 and passed to the Xsens SDK in https://github.com/robotology/wearables/blob/77362fc692bb9549b168b10c1200c473637bd59e/devices/XsensSuit/src/XsensSuit.cpp#L777 .

**Reply by lucaTagliapietra** [reply]
Yeap, I can confirm, you guys are talking about different SDK.

I don't think however it is discontinued, I think instead that XSens simply changed the name, now it is called MVN Studio Developer Toolkit. However, it is not free, it requires a "very" expensive license. Regarding the configurations, those match again the MVN Studio configuration you can choose, you will find more precise info also on the XSens MVN (Analyze or Animate is the same for this point) user manual.

**Reply by traversaro** [reply]
Thanks @lucaTagliapietra for chiming in!

**Comment by yeshasvitirupachuri** [comment]
> but probably you can find if from the old downloads on the Xsens website or asking to Xsens, probably @Yeshasvitvs or @lrapetti can know more about this. 

From the last time (September 2020) we spoke to a support representative at Xsens, it is not longer available from their website download section. You need to contact them to get the sdk. 

**UPDATE**

> Is the calibration (N/T pose and so on) performed by the Producer or the Consumer?

The calibration is done at the Producer layer, through yarp rpc port opened by [IXsensMVNControlWrapper](https://github.com/robotology/wearables/tree/master/wrappers/IXsensMVNControl) that implements [IXsensMVNControl](https://github.com/robotology/wearables/blob/master/interfaces/IXsensMVNControl/IXsensMVNControl.h) interface.

**Comment by torydebra** [comment]
Thanks for all the replies, I have not enough thumbs :+1: for everyone :D.

I may have found the _MVN Studio Developer Toolkit_ avaialbe to dowload. It is under the "tool" section of MVN Analyze/Anymate dowload box [here](https://www.xsens.com/software-downloads). It is an exe file that store some folders with c++ code. No readme nor some kind of docs is present, so I am not sure what it is, but it may be the SDK you are speaking about.

Following the chat we had yesterday:
I am looking at the HumanStateProvider code. [solveGlobalInverseKinematicsSolver()](https://github.com/robotology/human-dynamics-estimation/blob/5d12b7fb0cc63dffeecb4031425302a58fb37b31/devices/HumanStateProvider/HumanStateProvider.cpp#L1959-L2016) should be the "core" function that I am looking for. 

~~I am also waiting XSens response about possible sdk to combine only 6 IMUs to track only operator limbs.~~
XSens does not provide combination of data with custom number of devices. I will check the sdk of awinda to understand how data are gathered from the _single_ device

**Comment by yeshasvitirupachuri** [comment]
> It is under the "tool" section of MVN Analyze/Anymate dowload box https://www.xsens.com/software-downloads. 

@torydebra the link is not correct, please update it

> Following the chat we had yesterday: I am looking at the HumanStateProvider code. solveGlobalInverseKinematicsSolver() should be the "core" function that I am looking for.

@torydebra @lrapetti and @Yeshasvitvs had a quick chat yesterday to understand better what @torydebra is looking for. We gathered that @torydebra has a setup using https://github.com/qleonardolp/xsens_mtw_driver-release to get data from each individual xsens sensor, and is looking for a way to combine the sensor measurements to perform human arm tracking. So, we explained the aspects of inverse kinematics algorithms used for whole-body human motion tracking that are part of [HumanStateProvider](https://github.com/robotology/human-dynamics-estimation/tree/5d12b7fb0cc63dffeecb4031425302a58fb37b31/devices/HumanStateProvider), in particular the global ik based on [idyntree inverse kinematics](https://github.com/robotology/idyntree/tree/master/src/inverse-kinematics). 

> I am looking at the HumanStateProvider code. solveGlobalInverseKinematicsSolver()

@torydebra the link you pointed here corresponds to a commit, please be sure to use either the master or devel branch. 

> I am also waiting XSens response about possible sdk to combine only 6 IMUs to track only operator limbs.

As an alternative plan, we suggested to check with the xsens support to know how the hardware @torydebra has can be run with the latest sdk, that hopefully will also have example application code for configuration and data logging from xsens awinda system at hand.

**Reply by torydebra** [reply]
Yes, thanks for the recap for the other. 
I corrected the XSens link. The link of github is a permalink to the (actual) devel branch. I suppose github create it "pointing" to the commit in the case the branch changes and the link becomes meaningless

**Reply by torydebra** [reply]
I also update the message above because XSens answered to me

**Comment by torydebra** [comment]
Hi all,
Sorry for not updating this.

Your work is great, but it may be too extended for our simpler purposes. So we are going to different solutions, but we will keep updated on you work in any case for the future.

Thanks @Yeshasvitvs and all of you for the nice support!

**Comment by yeshasvitirupachuri** [comment]
Sure @torydebra, thanks for the update
Feel free to ping us if you see we can provide any support in the future
